gk/fib.c

```diff
@@ -1382,8 +1372,53 @@ check_prefix_locked(struct ip_prefix *prefix,
 	struct gk_fib *ip_prefix_fib;
 	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
 
-	if (action == GK_DROP || action == GK_FWD_GRANTOR)
+	if (action == GK_DROP || action == GK_FWD_GRANTOR) {
+		if (prefix->addr.proto == ETHER_TYPE_IPv4) {
+			uint8_t depth = prefix->len;
+			uint32_t next = 0;
+			int index = rte_lpm_rule_iterate(ltbl->lpm,
+				ntohl(prefix->addr.ip.v4.s_addr),
+				&depth, &next);
+			while (index >= 0) {
+				struct rte_lpm_rule *re =
+					&ltbl->lpm->rules_tbl[index];
+				ip_prefix_fib = &ltbl->fib_tbl[re->next_hop];
+				if (ip_prefix_fib->action != GK_FWD_GRANTOR &&
+						ip_prefix_fib->action !=
+						GK_DROP) {
+					RTE_LOG(WARNING, GATEKEEPER,
+						"gk: adding this rule with prefix %s and action %u would add a security hole since there already exists an entry of %u length with action %u",
+						prefix->str, action, depth, ip_prefix_fib->action);
+					return -1;
+				}
+				index = rte_lpm_rule_iterate(ltbl->lpm,
+					ntohl(prefix->addr.ip.v4.s_addr),
+					&depth, &next);
+			}
+		} else {
+			uint32_t next = 0;
+			int index = rte_lpm6_rule_iterate(ltbl->lpm6,
+				prefix->addr.ip.v6.s6_addr, prefix->len, &next);
```

Comment:
How come that `rte_lpm6_rule_iterate()` doesn't need the state variable `depht` as `rte_lpm_rule_iterate()` does?

Reply:
In IPv4, the LPM table is organized level by level, so we need the ``depth`` to indicate on which level we are iterating. However, in IPv6, all the rules are maintained in an array without order on depth. So, it's unnecessary to iterate the rules level by level.

Reply:
Okay. I recommend passing `depth` by reference in `rte_lpm6_rule_iterate()` as in `rte_lpm_rule_iterate()` to make the interface uniform, and hopefully future proof.

Also, assign `const` for the first two parameters of `rte_lpm6_rule_iterate()` and the first parameter of `rte_lpm_rule_iterate()`.